src/async/control-flow.md

```diff
@@ -0,0 +1,200 @@
+    let resp = reqwest::get(url).await?;
+    Ok(resp.text().await?.len())
+}
+
+#[tokio::main]
+async fn main() {
+    let urls: [&str; 4] = [
+        "https://google.com",
+        "https://httpbin.org/ip",
+        "https://play.rust-lang.org/",
+        "BAD_URL",
+    ];
+    let futures_iter = urls.into_iter().map(size_of_page);
+    let results = future::join_all(futures_iter).await;
+    let page_sizes_dict: HashMap<&str, Result<usize>> =
+        urls.into_iter().zip(results.into_iter()).collect();
+    println!("{:?}", page_sizes_dict);
+}
+```
+
+## Run multiple futures concurrently until the first one resolves: `select`
+
+### Equivalents:
+
+- JS: `Promise.race`
+- Python: `asyncio.new_event_loop().run_until_complete(asyncio.wait(task_set, return_when=asyncio.FIRST_COMPLETED))`
+
+```rust,editable
+use anyhow::Result;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::{sleep, Duration};
+
+#[derive(Debug, PartialEq)]
+enum Animal {
+    Cat { name: String },
+    Dog { name: String },
+}
+
+async fn first_animal_to_finish_race(
+    mut cat_rcv: Receiver<String>,
+    mut dog_rcv: Receiver<String>,
+) -> Option<Animal> {
+    tokio::select! {
+        cat_name = cat_rcv.recv() => Some(Animal::Cat { name: cat_name? }),
+        dog_name = dog_rcv.recv() => Some(Animal::Dog { name: dog_name? })
+    }
+}
+
+async fn simulate_animals_race(
+    sender: Sender<String>,
+    duration: Duration,
+    name: String,
+) -> Result<()> {
+    sleep(duration).await;
+    sender.send(name).await?;
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() {
+    let (cat_sender, cat_receiver) = mpsc::channel(32);
+    let (dog_sender, dog_receiver) = mpsc::channel(32);
+    tokio::spawn(simulate_animals_race(
+        cat_sender,
+        Duration::from_secs(10),
+        String::from("Felix"),
+    ));
+    tokio::spawn(simulate_animals_race(
+        dog_sender,
+        Duration::from_secs(5),
+        String::from("Rex"),
+    ));
+
+    let winner = first_animal_to_finish_race(cat_receiver, dog_receiver)
+        .await
+        .expect("Failed to receive winner");
+
+    println!("The winner of the race is {:?}", winner);
+
+    assert_eq!(
+        winner,
+        Animal::Dog {
+            name: String::from("Rex")
+        }
+    );
+}
+```
+
+## Iterate over multiple channels with a timeout: `select` with `pin`
+
+```rust,editable
+use anyhow::Result;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::{sleep, Duration};
+
+#[derive(Debug, PartialEq)]
+enum AnimalRacePerformance {
+    Cat { name: String },
+    Dog { name: String },
+}
+
+async fn race_finish_line(
+    mut cat_rcv: Receiver<String>,
+    mut dog_rcv: Receiver<String>,
+    timeout: Duration,
+) -> Option<Vec<AnimalRacePerformance>> {
+    let mut performances: Vec<AnimalRacePerformance> = Vec::new();
+    let timeout_sleep = sleep(timeout);
+    // See the chapter about pin.
+    tokio::pin!(timeout_sleep);
```

Comment:
Alternatively, that can be done with a oneshot channel, so we don't need to talk about pin

Reply:
Let's move this to the "pitfalls" section. AIUI the issue here is that you have a single Future that you'd like to `await` until it is Ready. This is different from the futures returned from `.recv()`, which can be created fresh every time around the loop. Maybe the example (in the pitfalls section) can start with
```rust
loop {
  select! {
    ..
    _ = sleep(timeout) => break,
  }
}
```
and then understand that this starts the timeout over for every iteration of the loop, and work out how to solve it from there.

I have to be honest, I don't think I could explain why `tokio::pin` is necessary here without a good bit of refreshing myself!